ioSPI/fourier.py

```diff
@@ -0,0 +1,237 @@
+        computes on the fly.
+    numpy_fft : func
+        Function backend for performing the fft.
+    only_real : bool
+        If True, return only real part of transform.
+
+    Returns
+    -------
+    arr3d_f : numpy.ndarray, shape (N,N,N)
+        Output array.
+    """
+    # compute on the fly if not precomputed
+    if neg_pos_3d is None:
+        neg_pos_3d = make_neg_pos_3d(arr3d.shape)
+        # extra multiplication for mod 4
+        if arr3d.shape[0] % 4 != 0:
+            neg_pos_3d *= -1
+
+    if mode == "forward":
+        arr3d_f = numpy_fft.fftn(neg_pos_3d * arr3d)
+        arr3d_f /= np.sqrt(np.prod(arr3d_f.shape))
+    elif mode == "inverse":
+        arr3d_f = numpy_fft.ifftn(neg_pos_3d * arr3d)
+        arr3d_f *= np.sqrt(np.prod(arr3d_f.shape))
+
+    if only_real:
+        arr3d_f = arr3d_f.real
+
+    arr3d_f *= neg_pos_3d
+    return arr3d_f
+
+
+def fft2d(
+    arr2d,
+    mode,
+    numpy_fft=pyfftw.interfaces.numpy_fft,
+    only_real=False,
+    batch=False,
+):
+    """2D FFT
+
+    Fast fourier transform. We apply an alterating +1/-1 multiplicative
+    before we go to/from Fourier space.
+    Later we apply this again to the transform.
+    The checkerboard pattern is applied instead of fft shifting
+    to have the dc component in centre of image
+    (even number of pixels, one to right of centre).
+
+    Parameters
+    ----------
+    arr2d : numpy.ndarray, shape (N,N) or (num_exposures,N,N)
+        Input array.
+    mode : str
+        Forward or reverse
+    numpy_fft : func
+        Function backend for performing the fft.
+    only_real : bool
+        If True, return only real part of transform.
+    batch : bool
+        Use for batch of exposures.
+
+    Returns
+    -------
+    arr2d_f : numpy.ndarray, shape (N,N) or (num_exposures,N,N)
+        Output array.
+    """
+    # needs work: look into pyfftw.interfaces.numpy_fft.irfftn .
+    # needs work: throw error instead of using assert
+    # assert (arr2d.ndim == 2 and not batch) or (batch and arr2d.ndim == 3)
+    n1, n2 = arr2d.shape[-2:]
+    # assert n1 == n2
+    # we apply an alterating +1/-1 multiplicative
+    # before we go to/from Fourier space.
+    # Later we apply this again to the transform.
+    # checkerboard pattern applied instead of fft shifting
+    # to have dc component in centre of image
+    # (even number of pixels, one to right of centre)
+
+    # reshape to 3 dimensions
+    arr2d = arr2d.reshape(-1, n1, n1)
+    neg_pos_2d = make_neg_pos_2d(arr2d.shape)
+    neg_pos_2d *= neg_pos_2d
+
+    if mode == "forward":
+        arr2d_f = numpy_fft.fftn(arr2d, axes=(-2, -1))
+        arr2d_f /= n1
+    elif mode == "inverse":
+        arr2d_f = numpy_fft.ifftn(arr2d, axes=(-2, -1))
+        arr2d_f *= n1
+
+    if only_real:
+        arr2d_f = arr2d_f.real
+
+    arr2d_f *= neg_pos_2d
+    if not batch:
+        arr2d_f = arr2d_f.reshape(n1, n2)
+
+    return arr2d_f
+
+
+def do_fft(arr, d=3, only_real=False, **kwargs):
```

Comment:
I would rename `d` in `dim` because:
- it is easier to understand (although I agree, this is not too complicated)
- when we refactor the code one day, it is easier to automatically change all the occurences of `dim` rather than all the occurences of `d`

Reply:
Please edit accordingly.

Reply:
Yes, good point. I'll try to get in the habit of being more verbose.